### PR TITLE
[WasmFS] Fix a bug and reduce contention on `close`

### PR DESCRIPTION
**Pull request by tlively**
We were previously calling the backend close method on files every time one of
their fds was closed, but we should have only been calling close on the files
when the last reference to one of their file descriptors was removed from the
file file table. Since the `dup` family of system calls copies fds by reference
rather than by value, these are not the same thing. Due to this bug, it was
possible for `close` to be called on a file more than `open` had been called on
it, which could put backends into invalid states and cause crashes. Fix the
problem by maintaining an explicit reference count on open file descriptors. Do
not just reuse the existing shared_ptr use_count to avoid accidentally
overcounting due to temporary copies of the shared_ptr that are not in the
table.

While fixing that bug, also take the opportunity to reduce lock contention due
to `close` by releasing the file table lock before calling into the backend to
close a file. `close` can be an expensive operation in the backend and the file
table has to be locked frequently, so this was previously a major source of
contention.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #17873** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/17873" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/17873?utm_source=stack-comment).

File: system/lib/wasmfs/syscalls.cpp [hunk 1]
```diff
@@ -1293,11 +1303,11 @@ int __syscall_pipe(intptr_t fd) {
   auto reader = std::make_shared<PipeFile>(S_IRUGO, data);
   auto writer = std::make_shared<PipeFile>(S_IWUGO, data);
 
-  auto fileTable = wasmFS.getFileTable().locked();
-
   std::shared_ptr<OpenFileState> openReader, openWriter;
   (void)OpenFileState::create(reader, O_RDONLY, openReader);
   (void)OpenFileState::create(writer, O_WRONLY, openWriter);
+
+  auto fileTable = wasmFS.getFileTable().locked();
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:1310]: This is a drive-by fix to also reduce lock contention in `pipe`, although this isn't nearly as important as the contention in `close`.

**Review by kripken — APPROVED** [review]
